core/network/reputation_change.hpp

```diff
@@ -0,0 +1,106 @@
+/**
+ * Copyright Soramitsu Co., Ltd. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#ifndef KAGOME_NETWORK_REPUTATIONCHANGE
+#define KAGOME_NETWORK_REPUTATIONCHANGE
+
+#include <chrono>
+
+#include <libp2p/peer/peer_id.hpp>
+
+namespace kagome::network {
+
+  using Reputation = std::int32_t;
+
+  struct ReputationChange {
+   public:
+    const Reputation value;
+    const std::string_view reason;
+
+    ReputationChange operator+(Reputation delta) const {
+      if (delta < 0) {
+        if (value < 0 and value + delta > 0) {  // negative overflow
+          return {std::numeric_limits<Reputation>::min(), reason};
+        }
+      } else if (delta > 0) {
+        if (value > 0 and value + delta < 0) {  // positive overflow
+          return {std::numeric_limits<Reputation>::max(), reason};
+        }
+      }
+      return {value + delta, reason};
+    }
+
+    ReputationChange operator*(size_t times) const {
+      auto limit = value < 0 ? std::numeric_limits<Reputation>::min()
+                             : std::numeric_limits<Reputation>::max();
+      if (static_cast<size_t>(limit / value) < times) {  // overflow
+        return {limit, reason};
+      }
+      return {static_cast<Reputation>(value * times), reason};
+    }
+  };
+
+  namespace reputation {
+
+    // clang-format off
+    namespace cost {
```

Comment:
just a suggestion, not a change requirement
```suggestion
    namespace penalty {
```

Reply:
it is taken from substrate